pkg/network/common/egressip.go

```diff
@@ -299,6 +374,124 @@ func (eit *EgressIPTracker) UpdateHostSubnetEgress(hs *osdnv1.HostSubnet) {
+		}
+	}
+}
+
+// validateEgressCIDRsAreSubnetOfCloudNetwork checks that whatever egress CIDRs are specified
+// on the HostSubnet also are a subnet of the cloud network. A failure to
+// specify this correctly might lead to egress IP assignments by the SDN which
+// are not allowed by the cloud provider.
+func (eit *EgressIPTracker) validateEgressCIDRsAreSubnetOfCloudNetwork(hs *osdnv1.HostSubnet) error {
+	cloudEgressIPConfig, err := eit.validateEgressIPConfigExists(hs)
+	if err != nil {
+		return err
+	}
+	_, cloudNetwork, _ := net.ParseCIDR(cloudEgressIPConfig.IFAddr.IPv4)
+	for _, egressCIDR := range hs.EgressCIDRs {
+		_, egressSubnet, _ := net.ParseCIDR(string(egressCIDR))
+		if !isSubnet(cloudNetwork, egressSubnet) {
+			return fmt.Errorf("EgressCIDR: %v is not a subnet of the cloud network: %v", egressSubnet, cloudNetwork)
+		}
+	}
+	return nil
+}
+
+// validateEgressIPCapacity checks that whatever egressIPs are specified
+// on the HostSubnet do not superseed the capacity of the node.
+func (eit *EgressIPTracker) validateEgressIPCapacity(hs *osdnv1.HostSubnet) error {
+	cloudEgressIPConfig, err := eit.validateEgressIPConfigExists(hs)
+	if err != nil {
+		return err
+	}
+	capacity := cloudEgressIPConfig.Capacity.IP
+	if capacity == 0 {
+		capacity = cloudEgressIPConfig.Capacity.IPv4
+	}
+	if len(hs.EgressIPs) > capacity {
+		return fmt.Errorf("the amount of requested EgressIPs (%v) on hostSubnet: %q superseeds the node's capacity (%v)", len(hs.EgressIPs), hs.Name, capacity)
+	}
+	return nil
+}
+
+func (eit *EgressIPTracker) validateEgressIPConfigExists(hs *osdnv1.HostSubnet) (*nodeCloudEgressIPConfiguration, error) {
+	node, err := eit.nodeInformer.Lister().Get(hs.Host)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving related node object, err: %v", err)
+	}
+	cloudEgressIPConfig, err := eit.getNodeCloudEgressIPConfig(node)
+	if err != nil {
+		return nil, err
+	}
+	if cloudEgressIPConfig == nil {
+		return nil, fmt.Errorf("related node object %q has an incomplete annotation %q, CloudEgressIPConfig: %+v", node.Name, nodeEgressIPConfigAnnotationKey, cloudEgressIPConfig)
+	}
+	return cloudEgressIPConfig, nil
+}
+
+func (eit *EgressIPTracker) getNodeCloudEgressIPConfig(node *corev1.Node) (*nodeCloudEgressIPConfiguration, error) {
+	nodeCloudEgressIPAnnotation, exists := node.Annotations[nodeEgressIPConfigAnnotationKey]
+	if !exists {
+		return nil, nil
+	}
+	cloudEgressIPConfig := []nodeCloudEgressIPConfiguration{}
+	if err := json.Unmarshal([]byte(nodeCloudEgressIPAnnotation), &cloudEgressIPConfig); err != nil {
+		return nil, fmt.Errorf("error de-serializing annotation: %q, err: %v", nodeCloudEgressIPAnnotation, err)
+	}
+	if len(cloudEgressIPConfig) > 0 {
+		return &cloudEgressIPConfig[0], nil
+	}
+	return nil, nil
+}
+
+func (eit *EgressIPTracker) initNodeCapacity(node *corev1.Node, nodeEgress *nodeEgress) error {
+	if nodeEgress.capacity == unlimitedNodeCapacity {
```

Comment:
don't put the whole function inside an if body:

    if nodeEgress.capacity != unlimitedNodeCapacity {
            return nil
    }
    ...